core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java

```diff
@@ -0,0 +1,181 @@
+
+    private static final int ACTIVE_SHARD_COUNT_DEFAULT = -2;
+    private static final int ALL_ACTIVE_SHARDS = -1;
+
+    public static final ActiveShardCount DEFAULT = new ActiveShardCount(ACTIVE_SHARD_COUNT_DEFAULT);
+    public static final ActiveShardCount ALL = new ActiveShardCount(ALL_ACTIVE_SHARDS);
+    public static final ActiveShardCount NONE = new ActiveShardCount(0);
+    public static final ActiveShardCount ONE = new ActiveShardCount(1);
+
+    private final int value;
+
+    private ActiveShardCount(final int value) {
+        this.value = value;
+    }
+
+    /**
+     * Get an ActiveShardCount instance for the given value.  The value is first validated to ensure
+     * it is a valid shard count and throws an IllegalArgumentException if validation fails.  Valid
+     * values are any non-negative number.  Directly use {@link ActiveShardCount#DEFAULT} for the
+     * default value (which is half the shards) or {@link ActiveShardCount#ALL} to specify all the shards.
+     */
+    public static ActiveShardCount from(final int value) {
+        if (value < 0) {
+            throw new IllegalArgumentException("shard count cannot be a negative value");
+        }
+        return get(value);
+    }
+
+    private static ActiveShardCount get(final int value) {
+        switch (validateValue(value)) {
+            case ACTIVE_SHARD_COUNT_DEFAULT:
+                return DEFAULT;
+            case ALL_ACTIVE_SHARDS:
+                return ALL;
+            case 0:
+                return NONE;
+            case 1:
+                return ONE;
+            default:
+                return new ActiveShardCount(value);
+        }
+    }
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeInt(value);
+    }
+
+    public static ActiveShardCount readFrom(final StreamInput in) throws IOException {
+        return get(in.readInt());
+    }
+
+    private static int validateValue(final int value) {
+        if (value < 0 && value != ACTIVE_SHARD_COUNT_DEFAULT && value != ALL_ACTIVE_SHARDS) {
+            throw new IllegalArgumentException("Invalid ActiveShardCount[" + value + "]");
+        }
+        return value;
+    }
+
+    /**
+     * Resolve this instance to an actual integer value for the number of active shard counts.
+     * If {@link ActiveShardCount#DEFAULT} or {@link ActiveShardCount#ALL} are specified, then
+     * the given {@link IndexShardRoutingTable} is used to determine the actual active shard count.
+     * The default value indicates half the shards must be active, but if there is only one shard
+     * in the index, then the value for the default will be 1 (instead of 0).
+     */
+    public int resolve(final IndexShardRoutingTable shardRoutingTable) {
+        Objects.requireNonNull(shardRoutingTable);
+        if (this == ActiveShardCount.DEFAULT) {
+            int val = shardRoutingTable.size() / 2; // default is half the shards must be active
```

Comment:
> Should this be (shardRoutingTable.size() + 1) / 2 with special case for shardRoutingTable.size == 2?

I think its debatable.  If `shardRoutingTable.size()` is 5, then we will return 3 instead of 2, which is more "quorum" than "half".  I don't mind if the semantics are "at least half".

> I wonder if we should just let the default be 1 instead of confusing the user with some sort of QUORUM semantics that does not exist in our replication model.

@bleskes suggested the default be "half" instead of 1.  I think if we view it as a "don't let a write proceed if we are down to just 1 shard", then maybe "half" is safer as a default, without introducing the word "quorum" or a notion of quorum, which may be confusing to the user.